blog/_posts/2019-03-02-custom-primitiv-type.md

```diff
@@ -0,0 +1,383 @@
+
+
+## Core.Intrinsics
+
+As we mentioned above, most of the function we need are lying under `Core.Intrinsics`. We will introduce some function we need. If you are familiar with LLVM Intrinsics, you will feel more at ease with these functions. (we will not include the floating-point part) 
+
+1.  Type construction
+    
+    1.  `Core.Intrinsics.bitcase(type, value)`
+    
+        `bitcase` told the LLVM to make a new same sized value of a primitive type instance `value`         and mark it as type `type`. The type of `value` must have the same bitsize as `type`.
+    
+    2.  `Core.Intrinsics.trunc_int(type, value)`
+    
+        `trunc_int` will truncate `value` to the bitsize of `type` and mark it as type `type`. The bitsize of `value` shoud be larger than `type`.
+    
+    3.  `Core.Intrinsics.sext_int(type, value)`
+    
+        `sext_int` do the sign extension on `value` and mark it as type `type`.        The bitsize of `value` shoud be smaller than `type`.
+    
+    4.  `Core.Intrinsics.zext_int(type, value)`
+    
+        `zext_int` do the zero extension on `value` and mark it as type `type`.
+    The bitsize of `value` shoud be smaller than `type`.
+
+2.  Arithmetic operation
+    1.  `Core.Intrinsics.add_int(value1, value2)`
+    
+        compute `value1 + value2`, as two integer addition.
+    
+    2.  `Core.Intrinsics.sub_int(value1, value2)`
+    
+        compute `value1 - value2`, as two integer subtraction.
+    
+    3.  `Core.Intrinsics.mul_int(value1, value2)`
+    
+        compute `value1 * value2`, as two integer multiplication.
+    
+see [Core.Intrinsics Reference](#reference) for more functions. 
+
+Now we have the basic function we need, we will start to define our own methds.
+
+
+## Define methods
+
+All the basic methods of our new primitve type should use those functions. For defining our own constructor, we need to construct from some built-in primitive 
+type. 
+
+For example: 
+
+```julia
+import Base: +
+primitive type A 72 end
+```
+
+Here is the constructor for `A`. extend a `Int64` to 72 bit with 0 and mark as type `A`.
+
+```julia
+A(x::Int64) = Core.Intrinsics.zext_int(A, x)
+```
+
+We can also accept and truncate a `Int128` to 72 bit and mark as type `A`.
+
+```julia
+A(x::Int128) = Core.Intrinsics.trunc_int(A, x)
+```
+
+`Base.+` is overloaded for defining addition on type `A`.
+
+```julia
++(x::A, y::A) = Core.Intrinsics.add_int(x, y)
+```
+
+Consequently, the usage of primitive type `A` could be looked like this:
+
+```julia
+a1 = A(3) #create instance of A
+a2 = A(12345678901234567890)
+
+a1 + a2 # add the value of a1 and a2
+```
+
+
+## Q&A
+
+Q. What's the benefit of using a custom type? Why not just define a wrapper type of type alias of a built-in primitive type?
+A. 
+First, if you want to dispatch on the type, than type alias might not work since it is actually the same type as the origin. 
```

Comment:
```suggestion
First, if you want to dispatch on the type, then a type alias will not work since it is actually the same type as the aliased one. 
```